Rule: Headers attribute only refers to cells in the same table (a25f45)

jkodu opened the pull request:
Rule: `headers` attribute only refers to `cells` in the same `table`

Closes issue: https://github.com/auto-wcag/auto-wcag/issues/406

Note:
For clear intents and purposes, the scope of this rule is kept to be fairly simple.
Intent is to author more rules for various other scenarios that involves this SC.

jkodu commented:
Note:

After discussion, it was decided to look at splitting the rule and or narrow the rule to one of the below expectations.

1) Check that cell that have a colspan or rowspan, have headers that exist in the table. (Automated)

2) In Complex tables, for cells that have headers referenced, ensure that the association of the headers is correct. (SEMI)

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells.md:
```diff
@@ -0,0 +1,208 @@
+
+### Applicability
+
+This rule applies to `td` and `th` elements with a `headers` attribute and within a `table` element, that is either [visible](#visible) or [included in the accessibility tree](#included-in-the-accessibility-tree).
```
"is" > "are"

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells.md:
```diff
@@ -0,0 +1,208 @@
+
+### Expectation
+
+Each target element with the `headers` attribute refers to other `cells` of the same `table` element.
```
I find using "refers to other cells" to be interpretable. What about something more clear like:
All `id`(s) listed in the `headers` attribute of the target element match the `id` of a `cell` in the same `table` element.

Also, the expectation leaves out the need for the cell with the matching id to be a header element. I suppose that's on purpose.

carlosapaduarte replied:
@WilcoFiers No, I don't know any UA that has issues with it. From that perspective, I don't have a problem with the expectation leaving out the need for the cell to be a header. I guess I'm just confused because sometimes we seem to try to follow the latest specs while other times we don't :-)

audreymaniez replied:
No I think  you're right @carlosapaduarte With nvda/ff for example, referencing td element in headers seems to not working, no headings are announced by the AT.

````html
 <table>
	<tr>
		<td id="one">my first header</td>
		<td id="two">second header</td>
	</tr>
	<tr>
		<td headers="two">cell</td>
		<td headers="one">other cell</td>
	</tr>
</table>
````

When focus on td `cell` or `other cell` no heading is announcing by the screen reader, that just says "cell" or "other cell".

Replacing td by th

````html
<th id="one">my first header</th>
<th id="two">second header</th>
````

and when the cell is focused, the AT announces the heading  : "my first header comun one my first header"

audreymaniez replied:
Plus, even it is not required to have a complete validation, it is a code validation error not to add id  that are not present on `th` element

![image](https://user-images.githubusercontent.com/409392/56261556-beff0880-60db-11e9-99a4-8340b528cf8e.png)

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells.md:
```diff
@@ -0,0 +1,208 @@
+
+## Assumptions
+
+- This test assumes that `headers` pointing to an `idref` of a non-existent element is a failure.
```
Is this really needed? If the id does not exist then it is not in the same table, right?

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells.md:
````diff
@@ -0,0 +1,208 @@
+</table>
+```
+
+#### Failed example 3
````
I don't think this tests something that hasn't been tested in the previous 2 failed examples

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells.md:
````diff
@@ -0,0 +1,208 @@
+</table>
+```
+
+### Inapplicable
````
You might want to add one inapplicable example with a table that is not visible neither included in the accessibility tree

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells.md:
```diff
@@ -0,0 +1,208 @@
+
+#### Inapplicable example 2
+
+A table used for presentation only, that has a `role="presentation"`. 
```
It's still visible. It's not applicable because it doesn't have any `headers` attribute.

jkodu replied:
@carlosapaduarte see the updated applicability.

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells.md:
```diff
@@ -0,0 +1,208 @@
+
+### Applicability
+
+This rule applies to `td` and `th` elements with a `headers` attribute and within a `table` element, that is either [visible](#visible) or [included in the accessibility tree](#included-in-the-accessibility-tree).
```
I don't think this should test cells that are excluded from the accessibility tree.

ShadowBB replied:
I agree. Cells that are excluded don't need correct associations as AT don't (or at least should not) communicate their relationships anyway.

ShadowBB commented on _rules/table-headers-attribute-refer-to-data-cells.md:
```diff
@@ -0,0 +1,208 @@
+
+### Expectation
+
+Each target element with the `headers` attribute refers to other `cells` of the same `table` element.
```
I don't think the table would fail WCAG if a header attribute refers to a non existing cell but there was no relationship to communicate anyway. What would be the harm in that WCAG wise?

Doesn't this rule need the same "the headers attribute is being used to comply to WCAG" assumption like all the other "spellchecking" rules?

audreymaniez replied:
I think that's the [F90](https://www.w3.org/WAI/WCAG21/Techniques/failures/F90) : Failure of Success Criterion 1.3.1 for incorrectly associating table headers and content via the headers and id attributes .

"The failure occurs when the relationship between data cells and corresponding header cells cannot be programmatically determined correctly because **the association of id and headers attributes is faulty.**"

There's a real issue on accessing the information if the relationship between heading and cell is invalid, because users that rely only on AT (screen reader) may access to wrong informations, and they do not have any other way to access the correct information (eg the visual presentation of the table for example). Those users may never find or find a wrong information if those relationships are wrong, and they will never know the information is wrong. 

(And I think we should having that F90 reference in the background section?)

ShadowBB commented on _rules/table-headers-attribute-refer-to-data-cells.md:
```diff
@@ -0,0 +1,208 @@
+
+### Expectation
+
+Each target element with the `headers` attribute refers to other `cells` of the same `table` element.
```
I also think the cell that is being refered to should be included in the accessibility tree right?

audreymaniez commented on _rules/table-headers-attribute-refer-to-data-cells.md:
```diff
@@ -0,0 +1,223 @@
+
+### Applicability
+
+This rule applies to `td` and `th` elements with a `headers` attribute and within a `table` element with a [semantic role](#semantic-role) of `table` or `grid` or `treegrid`, that is [included in the accessibility tree](#included-in-the-accessibility-tree).
```
If we scope `grid` and `treegrid` don't we need to scope to `columnheader` and `rowheader` too ? (https://www.w3.org/TR/wai-aria-practices-1.1/#treegrid_roles_states_props  - https://www.w3.org/TR/wai-aria-practices-1.1/#grid_roles_states_props). 

Can we just focus on role `table` ?

Do we need to specify too that th needs to not have an explicit role define ? td and th with no explicit semantic role or the same as the implicit role  ?

> This rule applies to `td` and `th` elements with a `headers` attribute and within a `table` element with a [semantic role](#semantic-role) of `table` that is [included in the accessibility tree](#included-in-the-accessibility-tree). `td` and `th` elements has no explicit semantic role or an explicit semantic role that is identical to their explicit semantic role.

carlosapaduarte replied:
I agree with @audreymaniez suggestions. Just some corrections to the last sentence of the proposed applicability:
`td` and `th` elements have no explicit semantic role or an explicit semantic role that is identical to their implicit semantic role.

annethyme replied:
We discussed a little internally in our office, and would suggest that we in this rule look at the semantic roles of elements (table, grid, rowheader, columnheader, cell), instead of the html element names (table, th, td).
This will be much in line with most other rules we have written.

And then, like we have discussed previously, we would suggest that we don't limit the scope of the rule to cells that have a `headers` attribute, but in general just check whether a header can be calculated for that particular cell (there is an algorithm for it in the html spec: https://www.w3.org/TR/html/tabular-data.html#forming-relationships-between-data-cells-and-header-cells).

@kasperisager also had the point that if we want to catch the roles `table`, `grid` and `treegrid`, you can just say something like "with the semantic role of `table` or a semantic role that inherits from the `table` role".
Then if you really want to list the roles too, you can have a note saying: "The roles that inherit from the `table` role are `grid` and `treegrid`". 
But in this way, it will be clear what those particular roles have in common and why they are in this rule together.

jkodu replied:
Based on the discussion with @WilcoFiers & @annethyme, we concluded that we will scope this to only `table` elements. I have made some updates to the rule.

audreymaniez commented on _rules/table-headers-attribute-refer-to-data-cells.md:
```diff
@@ -0,0 +1,223 @@
+
+- [Understanding Success Criterion 1.3.1: Information and relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html)
+- [H43: Using id and headers attributes to associate data cells with header cells in data tables](https://www.w3.org/WAI/WCAG21/Techniques/html/H43)
+- [Failure of Success Criterion 1.3.1 for incorrectly associating table headers and content via the headers and id attributes](https://www.w3.org/WAI/WCAG21/Techniques/failures/F90)
```
I think we should identify the failure by its number (as it is done with the technic above ?) 

> - [F90: Incorrectly associating table headers and content via the headers and id attributes](https://www.w3.org/WAI/WCAG21/Techniques/failures/F90)

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells.md:
```diff
@@ -0,0 +1,223 @@
+
+### Expectation
+
+Each target element with the `headers` attribute refers to other `th` of the same `table` element.
```
Should we include cells with the columnheader role in addition to `th` elements?

annethyme replied:
If it was a `td` with an id, would referring to that also work? 
I am thinking if it's enough that it's just another cell in the same table...

jkodu replied:
If the `td` that is being referred to, is marked as `row/columnheader` using a `role`.

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells.md:
```diff
@@ -0,0 +1,223 @@
+
+#### Inapplicable example 3
+
+The rule only applies only to `table` element.
```
Remove the first only

annethyme commented:
@jkodu, Siteimprove's alternative proposal for how to scope this rule to keep it in line with the other ACT rule that we are writing, that are based on semantics rather than HTML element types: 

## Applicability: 

This rule applies to any element with the [semantic role](#semantic-role) `cell` or `gridcell` that is [included in the accessibility tree](#included-in-the-accessibility-tree) and is the descendant (in the flat tree?) of an element that
- has the semantic role of table or a semantic role that inherits from the table role; 
- is [included in the accessibility tree](#included-in-the-accessibility-tree); and
- is not a [layout table](#layout-table) (DEFINITION OF LAYOUT TABLE NEEDED).

**Note:** The roles that inherit from the `table` role are `grid` and `treegrid`.

## Expectation: 
For the target element a [header cell is assigned](https://www.w3.org/TR/html/tabular-data.html#forming-relationships-between-data-cells-and-header-cells).

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,264 @@
+
+## Applicability
+
+This rule applies to `td` elements with a `headers` attribute, that is [included in the accessibility tree](#included-in-the-accessibility-tree).
```
is > are

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,264 @@
+
+#### Inapplicable Example 2
+
+A table used for presentation only, that has a `role="presentation"`. 
```
Shouldn't the main reason be that there are no `headers` attributes?

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,264 @@
+
+#### Inapplicable Example 3
+
+The rule applies only to `table` element.
```
`table` > `td`

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,264 @@
+
+#### Inapplicable Example 5
+
+A table that is not visible. 
```
There is nothing in the applicability about visibility. Not included in the accessibility is covered in the next test case, so you don't need to duplicate.

WilcoFiers commented:
@annethyme I disagree with you that what you are proposing as an alternative for this rule. Take a table like this:

```html
<table>
 <tr>
    <td></td>
    <th id="class1">1st class</th>
    <th id="class2">2nd class</th>
  </tr> <tr>
    <th id="r1">One way</th>
    <td headers="r1 cls1">12,40</td>
    <td headers="r1 cls2">8,00</td>
  </tr> <tr>
    <th id="r2">Return</th>
    <td headers="r2 cls1">21,90</td>
    <td headers="r2 cls2">14,10</td>
  </tr>
</table>
```

The mismatch between the `cls1` and `class1` ID here is a problem, despite the `<td>` having a heading. What you are proposing, while it certainly has its used, wouldn't catch the kind of issues that this rule is designed to capture; mistyped and/or incorrect use of IDs. Unlike lots of ARIA stuff, where missing IDREFs can be caused by dynamic, what's unique in the `headers` attribute is that whenever it's wrong, it's almost always an accessibility problem.

jkodu commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,247 @@
+
+## Expectation
+
+Each target element with the `headers` attribute refers to other `th` within the same `table` element.
```
@carlosapaduarte 
I believe we author most expectations to begin with `Each target element...`.
And the applicability can specify to choose a collection of element.

The language is such that of the many elements that apply to this rule, ensure that each of them satisfy something....

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,247 @@
+
+#### Inapplicable Example 4
+
+No `headers` attribute is defined for table `cell`.
```
This is the same as Inapplicable Example 1

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,247 @@
+
+## Expectation
+
+Each target element with the `headers` attribute refers to other `th` elements or cells with a [semantic role](#semantic-role) of `columnheader` or `rowheader` within the same `table` element.
```
I still think that the text of the Expectation could be improved. What about this:
Each target element refers, in its `headers` attribute, to other `th` elements or cells with a [semantic role](#semantic-role) of `columnheader` or `rowheader` within the same `table` element.

audreymaniez commented on pages/glossary/layout-table.md:
```diff
@@ -0,0 +1,14 @@
+
+Table that have no tabular mark-up are interpreted by assistive technologies as layout table. According to WCAG this includes:	
+
+- No header cells: that is `th` elements, or `td` with ARIA roles rowheader and columnheader.	
```
add quotes to role

> No header cells: that is `th` elements, or `td` with ARIA roles `rowheader` and `columnheader`.

audreymaniez commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,231 @@
+<div role="table">
+  <div role="row">
+    <div class="columnheader" id="header1">Projects</div>
+    <div class="columnheader" id="header2">Exams</div>
```
I suppose you meant `role="columnheader"`not `class`

````html

<div role="columnheader" id="header1">Projects</div>
<div role="columnheader" id="header2">Exams</div>

jkodu commented:
Final Call ends 30th July 2019

EmmaJP commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
````diff
@@ -0,0 +1,231 @@
+    <td headers="header3">15%</td>
+  </tr>
+</table>
+```
````
I think this example would also fail an HTML validation test. Is this still something to include in an accessibility test? (just double checking)

jkodu replied:
@EmmaJP in this case we are simply using a `header3` as an `id` that does not exist. I ran the HTML validation [`html-hint`](https://www.npmjs.com/package/htmlhint) against the above snippet and it seemed ok.

EmmaJP replied:
W3 nu validator throws an error when 'header 3' does not exist, and also when 'header 3' exists as an id in a different table. So this example is also invalid HTML.

I've always thought valid HTML a prerequisite for accessibility rules.

jkodu replied:
Right. I think we will need to fix this out of scope of this PR for 2 reasons.

- I currently have a different pr, which checks validity of HTML using `html-hint`, see https://github.com/act-rules/act-rules.github.io/pull/631
- I would like to integrate something like w3 nu validator as a part of continuous integration testing.

In this specific example we have refer to an id that does not exist, so yes it will refer to an `does not exist` error.

I have created an issue for this Emma - https://github.com/act-rules/act-rules.github.io/issues/679

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,231 @@
+
+## Expectation
+
+Each target element refers in its `headers` attribute to other `th` elements or cells with a [semantic role](#semantic-role) of `columnheader` or `rowheader` within the same `table` element.
```
I don't really think this expectation is accurate enough. What does "refers in its headers attribute" mean? What is a "cell", when is something in the "same table"? I've tried to write this out myself and came up with the following:

> The `header` attribute of each target element is [a set of space separated IDs]([set of space tokens](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens)), each of which is an ID of an element that:
> 
> 1. has a [semantic role](#semantic-role) of `columnheader` or `rowheader`, and
> 2. is a `cell` of the same [`table`](#https://www.w3.org/TR/html50/tabular-data.html#concept-table) as the target element

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,231 @@
+## Assumptions
+
+- This test assumes that the `headers` attribute is only used to identify table headers. If other information is included in the `headers` attribute, the rule may fail on issues that are not accessibility concerns. For example, if `headers` is used to include information for script, this rule may not be accurate.
+- Tables using `headers` id associations are intended as data rather than [layout table](#layout-table).
```
I don't think you need to define layout table here. Linking out to something about layout tables in WCAG is enough. We may need that term for some other rule.

jkodu replied:
Agreed with @WilcoFiers, to leave this as is for now.

jkodu commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,231 @@
+  wcag20:1.3.1: # Info and Relationships (A)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
```
There was never a test for this. Test only checks for top level frontmatter keys. I think we should add a test for this.

jkodu replied:
Added a test - PR https://github.com/act-rules/act-rules.github.io/pull/691

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,234 @@
+
+## Expectation
+
+The `header` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element that:
```
This seems more like a definition of `header` attribute than an expectation. This is my attempt at improving it:

Each ID in the [set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens) that is the `headers` attribute of the target element, is the ID of an element that:

jkodu replied:
Sorry Carlos, I feel like what is currently in place is straightforward over your suggestion.

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,234 @@
+
+The `header` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element that:
+1. has a [semantic role](#semantic-role) of `columnheader` or `rowheader`, and
+2. is a `cell` of the same [`table`](https://www.w3.org/TR/html50/tabular-data.html#concept-table) as the target element
```
I don't think cell should be in backticks.
But you might want to link to the definition: [cell](https://www.w3.org/TR/html52/tabular-data.html#cell)

jkodu commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
````diff
@@ -0,0 +1,234 @@
+</table>
+```
+
+#### Failed Example 2
````
Updated example 1 to have `NOT-EXIST` as the header id, but keeping fail example 2 as it is slightly different from example 1, given only one of the header id referenced is missing.

Jym77 replied:
Make the description more of the example more precise:  some headers are OK, 1 is KO and 1 KO is enough.
(this is different from other handlings of list in attributes where the list means "use the first valid one", here the list means "use all of them and they must all be valid")

annethyme commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,285 @@
+
+## Expectation
+
+The `header` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element that:
```
Shouldn't it at least be "The value of the `headers` attribute..." ?

jkodu replied:
we say things like the role attribute elsewhere, so I think this is conveyed by default.

annethyme commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,285 @@
+
+- This test assumes that the `headers` attribute is only used to identify table headers. If other information is included in the `headers` attribute, the rule may fail on issues that are not accessibility concerns. For example, if `headers` is used to include information for script, this rule may not be accurate.
+- Tables using `headers` id associations are intended as data rather than [layout table](#layout-table).
+
```
Should there be another assumption that if the IDs do not live up to the expectations in this rule, the browser will try to calculate a default header from the [algorithm for assigning header cells](https://html.spec.whatwg.org/multipage/tables.html#header-and-data-cell-semantics), and if the table is a simple table, then header cells might be correctly assigned even though the outcome of this rule is `failed` ?

jkodu replied:
Adding this will rise more questions like:
- what is do not live up to the expectations?
- what is a `simple table` etc.,

jkodu commented on pages/glossary/layout-table.md:
```diff
@@ -0,0 +1,14 @@
+- No (or empty) `summary` attribute.	
+- No `scope` or `headers` attributes on table cells.	
+- Table with single column or row, or table with less than two rows containing two valid data cells are automatically marked as layout table by assistive technologies. A cell is considered valid:	
+  - if its size can be obtained and is between 200 and 16000 pixels, or	
```
I believe `size` is used as a term that fits either height or width, based on whether they are a single column or single row. I believe it is self explanatory.

jkodu commented on pages/glossary/layout-table.md:
```diff
@@ -0,0 +1,14 @@
+- No `scope` or `headers` attributes on table cells.	
+- Table with single column or row, or table with less than two rows containing two valid data cells are automatically marked as layout table by assistive technologies. A cell is considered valid:	
+  - if its size can be obtained and is between 200 and 16000 pixels, or	
+  - if its size can't be obtained and it contains text (Cells only containing graphics are not considered valid).
```
At the moment I believe so.

jkodu commented:
@annethyme I have updated with the changes requested, as for the definition of the layout table, given it is only used in the assumption, enhancing the same would be a different issue as against affecting this PR.

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,287 @@
+name: Headers attribute of td element refers to cells in the same table element
+rule_type: atomic
+description: |
+  This rule checks that the headers attribute of `td` elements refer to cells in the same `table` element that has a semantic role of `columnheader` or `rowheader`.
```
cells ... that has a semantic role > cells ... that have a semantic role

carlosapaduarte commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,287 @@
+
+## Applicability
+
+This rule applies to any `td` element with a `headers` attribute, that are [included in the accessibility tree](#included-in-the-accessibility-tree).
```
that are > that is

carlosapaduarte commented on pages/glossary/layout-table.md:
```diff
@@ -0,0 +1,14 @@
+key: layout-table	
+---	
+
+Table that have no tabular mark-up are interpreted by assistive technologies as layout table. According to WCAG this includes, but is not limited to:	
```
Table > Tables
as layout table > as layout tables

carlosapaduarte replied:
I think we could use a definition of "tabular mark-up"

carlosapaduarte commented on pages/glossary/layout-table.md:
```diff
@@ -0,0 +1,14 @@
+
+Table that have no tabular mark-up are interpreted by assistive technologies as layout table. According to WCAG this includes, but is not limited to:	
+
+- No header [`cells`](https://www.w3.org/TR/html52/tabular-data.html#cell), that is [`th`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/th) elements, or [`td`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/td) elements with ARIA roles [`rowheader`](https://www.w3.org/TR/wai-aria-1.1/#rowheader) and [`columnheader`](https://www.w3.org/TR/wai-aria-1.1/#columnheader).	
```
that is > that are
rowheader and columnheader > rowheader or columnheader

carlosapaduarte commented on pages/glossary/layout-table.md:
```diff
@@ -0,0 +1,14 @@
+- No [`caption`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/caption) element.	
+- No (or empty) `summary` attribute.	
+- No `scope` or `headers` attribute on table cells.	
+- Table with single column or row, or table with less than two rows containing two valid data cells are automatically marked as layout table by assistive technologies. A cell is considered valid:	
```
Is it less "that two rows" or "two or less rows"? Because the first is equal to table with one row

jkodu commented:
@carlosapaduarte I have done the editorial changes. Thanks for the suggestions.

I have changed my mind about the layout table definition. Given it is used in an assumption, which I have now removed as it was a very obvious assumption. 

If we really need it, I would rather like to revive the [is data table](https://github.com/act-rules/act-rules.github.io/blob/develop/__deprecated__/glossary/outdated/is-data-table.md) definition instead of trying to describe layout table, which can then lead to a need of defining more terms out of them, and thus steer us to a never ending cyclic loop of definitions.

EmmaJP commented:
Not sure how helpful it will be, or where it was sourced from, but a predecessor include an algorithm for a data table in our older Accessible HTML guidance: http://www.bbc.co.uk/guidelines/futuremedia/accessibility/html/tables.shtml

kasperisager commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,286 @@
+
+## Applicability
+
+This rule applies to any `td` element with a `headers` attribute, that is [included in the accessibility tree](#included-in-the-accessibility-tree).
```
Should the rule not apply to the `headers` attribute instead of the `td` element?

jkodu replied:
I did some testing on VoiceOver in general `headers` attributes seems to have a lot of inconsistencies. Also tested Firefox + nvda combo, with help of testing team.

Some findings:
- td scope=row does NOT work with VO or NVDA
- using headers="id1 id2" does NOT work 
- using headers to point to row headers in 2nd column does not work in VO, only works in NVDA/Firefox.
- `th` with `headers` attribute were also inconsistent.

Given the inconsistencies, I think it is best to scope this rule to `td` elements for this rule.

kasperisager replied:
All good then, though my question still stands: Should the test target be the `headers` attribute instead of the `td` element? This would be consistent with other rules that solely target attributes.

kasperisager commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,286 @@
+
+## Expectation
+
+The `header` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element that:
```
The expectation needs to account for the presence of nested browsing contexts and Shadow DOM:

> ...each of which is an ID of an element **in the same document tree or shadow tree as the applicable element**.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/act-rules/act-rules.github.io?pullRequest=412) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/act-rules/act-rules.github.io?pullRequest=412) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/act-rules/act-rules.github.io?pullRequest=412) it.</sub>

kasperisager commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,287 @@
+
+## Applicability
+
+This rule applies to any [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) with a [`headers`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/td#attr-headers) attribute, in a [table](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table) element that is [included in the accessibility tree](#included-in-the-accessibility-tree).
```
Still not addressed: https://github.com/act-rules/act-rules.github.io/pull/412#discussion_r311592701. It could look like this:

> This rule applies to any [`headers`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/td#attr-headers) attribute specified on a [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) in a [table](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table) element that is [included in the accessibility tree](#included-in-the-accessibility-tree).

jkodu replied:
Good play in words. Will switcheroo.

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,287 @@
+
+## Applicability
+
+This rule applies to any [`headers`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/td#attr-headers) attribute specified on a [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) in a [table](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table) element that is [included in the accessibility tree](#included-in-the-accessibility-tree).
```
I don't think you should link the headers attribute or the table element. We're not doing that with element / attribute names in other rules. Also, MDN links don't include definitions, and those pages aren't stable.

WilcoFiers replied:
This rule probably shouldn't be applicable to `table` elements with a semantic role of `table` or `grid`. If the semantic role of table is changed, the heading assignment no longer applies. I'm not sure if this happens consistently in AT, so that might be worth testing.

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,287 @@
+---
+id: a25f45
+name: Headers attribute of td element refers to cells in the same table element
```
The applicability uses "cell" as a definition, rather than `td`. A cell, according to spec, can be either a `td` or a `th`, so the applicability is inconsistent with the rule name and description. In my opinion this rule shouldn't be limited to `td` elements, so I think the applicability is correct, but the name / description aren't, and there are missing test cases for `th` with a `headers` attribute.

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,287 @@
+
+## Expectation
+
+The `headers` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element in the same [document tree](https://www.w3.org/TR/dom41/#document-trees) or [shadow tree](https://www.w3.org/TR/dom41/#shadow-trees) as the applicable elment, that:
```
You don't need to look at trees here. A cell in the same table (the second bullet) can't be in a different tree. This part should simply be:

> The `headers` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element that:

kasperisager replied:
Allow me to abuse the DOM yet again:

```html
<table>
  <th>
    <div></div>
  </th>
  <td headers="foo"></td>
</table>

<script>
  const th = document.createElement("th");
  th.id = "foo";

  document
    .querySelector("div")
    .attachShadow({ mode: "open" })
    .appendChild(th);
</script>
```

jkodu replied:
@WilcoFiers seeing this comment - https://github.com/act-rules/act-rules.github.io/pull/412#discussion_r318053483, it think it is best to emphasise document/ shadow tree. Wdyt?

WilcoFiers replied:
@kasperisager that `th` element isn't in the same tree as the td (see the definitions of tree)

kasperisager replied:
@WilcoFiers Bah, I copy pasted the wrong code:

```html
<table>
  <th>
    <div></div>
  </th>
  <td headers="foo"></td>
</table>

<script>
  const th = document.createElement("th");
  th.id = "foo";

  document
    .querySelector("div")
    .appendChild(th);
</script>
```

jkodu replied:
Leaving usages of `document tree` and `shadow tree`.

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,287 @@
+## Assumptions
+
+- This test assumes that the `headers` attribute is only used to identify table headers. If other information is included in the `headers` attribute, the rule may fail on issues that are not accessibility concerns. For example, if `headers` is used to include information for script, this rule may not be accurate.
+- This test assumes that if the IDs of the `headers` attribute, do not live up to the expectations in this rule, the browser will try to calculate a default header from the [algorithm for assigning header cells](https://html.spec.whatwg.org/multipage/tables.html#header-and-data-cell-semantics).
```
I think it's actually the opposite. We're assuming that if `headers` attribute fails, the browser won't be able to come up with the correct headers. Also, use "expectation" rather than "expectations", since there's only one of them.

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,287 @@
+
+The `headers` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element in the same [document tree](https://www.w3.org/TR/dom41/#document-trees) or [shadow tree](https://www.w3.org/TR/dom41/#shadow-trees) as the applicable elment, that:
+
+1. has a [semantic role](#semantic-role) of `columnheader` or `rowheader`, and
```
I don't really understand why this is required. I know the spec says that you should only reference other headers with the `headers` attribute, but are there any accessibility issues that come from referencing a `td` element with the `headers` attribute? I don't see what they would be, unless maybe assistive tech don't announce `td`s if they're used in the `headers` attribute, but I don't think that's the case.

jkodu replied:
Note to self:
- Get data of AT testing.

jkodu replied:
@WilcoFiers 

I did some testing & other `td` referred via `headers` attribute is not consistently announced without a role of `columnheder` or `rowheader`. These tests were done with voiceover.

Note voiceover seems to be very inconsistent even when `th` is referred in `headers` attribute. It also seems like headers are only rightly announced when wrapped with `<thead>` in the case of voiceover.

Here is also an article on elaborate testing outcomes:
- [VoiceOver and Tables](http://accessibleculture.org/articles/2010/10/voiceover-and-tables-with-an-empty-first-header-cell/)

I did not test all of the test cases listed in the above example.

Note though in terms of automated implementation of this rule, I do think it does not matter whether a header attribute refers to `th` or `td`.

jkodu replied:
Did more testing referencing `td` with `headers` attribute, does not get announced in VoiceOver :/. Overall headers attribute is quite inconsistent with AT's.

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
````diff
@@ -0,0 +1,287 @@
+</table>
+```
+
+### Failed
````
Out of all of these, I think only failed example 3 actually fails SC 1.3.1. For the sake of having all failed examples be true failures of the SC, I think these things are going to need some work.

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
````diff
@@ -0,0 +1,287 @@
+</table>
+```
+
+### Failed
````
There is no test for when `headers` refers to a data call. I don't think that should fail this rule, so if that is changed this comment is resolved as well.

jkodu replied:
This feels like something an implementation should consider to return an outcome for these test cases, over a change in the rule?

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
````diff
@@ -0,0 +1,287 @@
+```html
+<table role="presentation">
+  <tr>
+    <th id="header1">Project Status</th>
````
Putting a `th` in a layout table may be controversial. I don't think it fails WCAG, but I know not everyone agrees with that. Putting this example in the way it is now means this rule is testing something that it's not about. I'd suggest changing the `th` to a `td` here.

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,287 @@
+
+#### Inapplicable Example 4
+
+A table that is not included in the accessibility tree. 
```
This would be inapplicable even without aria-hidden. Please add the `headers` attribute.

Jym77 commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,287 @@
+
+## Expectation
+
+The `headers` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element in the same [document tree](https://www.w3.org/TR/dom41/#document-trees) or [shadow tree](https://www.w3.org/TR/dom41/#shadow-trees) as the applicable elment, that:
```
Editorial: elment => element (end of the sentence)

Jym77 commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,287 @@
+
+## Expectation
+
+The `headers` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element in the same [document tree](https://www.w3.org/TR/dom41/#document-trees) or [shadow tree](https://www.w3.org/TR/dom41/#shadow-trees) as the applicable elment, that:
```
What happen if there are several element with that id?
```html
<table>
  <thead>	
    <tr>
      <th id="header1">Projects</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td headers="header1">15%</td>
    </tr>
<tr>
      <td id="header1">25%</td>
</tr>
  </tbody>
</table>
```

jkodu replied:
This could be more developer error than an actual scenario. I did VoiceOver testing and in some cases one is announced or none :/. I am unsure if this is within the scope of this rule.

Jym77 replied:
I agree that this would be more of a global HTML error than a problem with this rule.
Maybe it's worth an assumption "This rule assume that `id` are unique within the current document" to ensure that we can write "the ID of **an** element" without any ambiguity.

Jym77 commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,302 @@
+
+#### Failed Example 3
+
+Cells with `headers` attribute, refers to multiple non-existing elements within the same `table`.
```
```suggestion
Both `td` elements have `headers` attribute which refers to ID that do not exist within the same `table`.
```

Jym77 commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,302 @@
+
+#### Failed Example 4
+
+The `headers` attribute on the cell refers to an element inside the same `table`, but is not a `rowheader` or `columnheader`.
```
```suggestion
The `headers` attribute on the cell refers to an element inside the same `table` which does not have a role of `rowheader` or `columnheader`.
```

Jym77 commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,302 @@
+
+#### Inapplicable Example 1
+
+There's no usage of the `headers` attribute.
```
```suggestion
There is no `headers` attribute.
```

Jym77 commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,302 @@
+
+#### Inapplicable Example 2
+
+A table used for presentation only, that has a `role="presentation"`.
```
```suggestion
The `table` has a `role="presentation"` and thus is not [included in the accessibility tree](#included-in-the-accessibility-tree).
```

Jym77 commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,302 @@
+
+#### Inapplicable Example 3
+
+The rule applies only to `table` element.
```
```suggestion
The rule applies only to `headers` attribute within a `table` element.
```

Jym77 commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,302 @@
+
+#### Inapplicable Example 4
+
+A table that is not included in the accessibility tree.
```
```suggestion
The `table` is not [included in the accessibility tree](#included-in-the-accessibility-tree).
```

Jym77 commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
````diff
@@ -0,0 +1,302 @@
+</table>
+```
+
+### Inapplicable
````
Need an example with a `table` that is not visible (but included in the accessibility tree).

ShadowBB commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,302 @@
+## Assumptions
+
+- This test assumes that the `headers` attribute is only used to identify table headers. If other information is included in the `headers` attribute, the rule may fail on issues that are not accessibility concerns. For example, if `headers` is used to include information for script, this rule may not be accurate.
+- This test assumes that the `headers` attribute is required to express the relationship between data and header cells in the `table`. If the browser [computes an adequate fallback header](https://html.spec.whatwg.org/multipage/tables.html#header-and-data-cell-semantics), this rule may produce incorrect results.
```
The assumption is there because sometimes the relationship is expressed in different ways or sometimes there is no relationship at all and the header attribute is added in error (but also has a non existing ID). This assumption covers those situations.

jkodu commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,302 @@
+
+## Applicability
+
+This rule applies to any `headers` attribute specified on a [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) in a `table` element that is is [visible](#visible) and [included in the accessibility tree](#included-in-the-accessibility-tree).
```
I am unsure about hidden cells within a table & the scope of this rule to cater to them. Can other reviewers voice their opinions?

Jym77 commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
````diff
@@ -0,0 +1,318 @@
+		<td>15%</td>
+	</tr>
+</table>
+```
````
This is the same example as inapplicable example 2 where the table was **not** included in the accessibility tree…

Jym77 commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,318 @@
+
+## Applicability
+
+This rule applies to any `headers` attribute specified on a [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) within a `table` element that is is [visible](#visible) and [included in the accessibility tree](#included-in-the-accessibility-tree).
```
Even without considering invisible cells, you can still rephrase that to make it obvious that the "that is visible and included in the accessibility tree" part refers to the `table`, not the `cell`.

Jym77 commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,318 @@
+
+#### Passed Example 4
+
+This `table` has multiple `columnheader(s)`. The `headers` attribute on the cells lists the `id(s)` of `th` elements within the same `table`.
```
```suggestion
This `table` has multiple elements with a role of `columnheader`. The `headers` attribute on the cells lists IDs of `th` elements within the same `table`.
```

Jym77 commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
````diff
@@ -0,0 +1,318 @@
+</table>
+```
+
+#### Failed Example 3
````
This is quite redundant with Failed Example 2, but simplified. Could be removed.

Jym77 commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,318 @@
+
+#### Inapplicable Example 3
+
+The `table` is not [visible](#visible).
```
Rather put the table off-screen to make it not visible but still included in the accessibility tree.

Jym77 commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
````diff
@@ -0,0 +1,318 @@
+</div>
+```
+
+#### Inapplicable Example 5
````
Essentially the same as Inapplicable example 2, and we're not going to list all the ways a table can be excluded from the accessibility tree. Can be removed.

jkodu commented:
Final call ends **December 10, 2019**

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,295 @@
+
+## Accessibility Support
+
+- `headers` attribute may not be consistently announced by assistive technologies.
```
This is too much of a generalisation?

```suggestion
The `headers` attribute is not supported by some popular assistive technologies as a way to assign headers in a table. Passing this rule can still cause issues for users of those assistive technologies.
```

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,295 @@
+		<th id="headerOfColumn">Projects</th>
+	</tr>
+	<tr>
+		<td headers="elmOutsideTable">15%</td>
```
This comment was never addressed: https://github.com/act-rules/act-rules.github.io/pull/412#discussion_r318026911 Like I said before, I don't think failed example 1 and 2 actually fail the success criterion.

EmmaJP replied:
I disagree, sort of. Reading 1.3.1 is seems to be about the programatic information conveying the same association that is conveyed visually.

Failed examples 1 and 2 certainly match the rule expectations.

For failed example 1, it could be more clear that the span does not visually head up the table cell. For failed example 2, it could be that the Id doesn't exist in the page at all.

It strikes me that both examples are maybe an `aria-labelledby` away from not being a fail example. A point of view that raises questions about the rule and the referenced SC technique possibly not being detailed enough.

WilcoFiers replied:
This needs a little research.

Jym77 replied:
As per the [Algorithm for assigning header cells](https://html.spec.whatwg.org/multipage/tables.html#algorithm-for-assigning-header-cells), the position of the cell is only looked at if there is no `headers` attribute.

So, I think that in this case there is no programmatic association. (maybe UAs and ATs are fallbacking to the position if the `headers` attribute is KO...)

WilcoFiers replied:
Every screen reader we've tested this in will look at row/column headers for fallback. That may not be what the algorithm says they should do, but someone testing with one of the top 5 AT / browser combinations is not going to see a problem here. I find this test case questionable. Here's the test case we ran a little while ago:

```html
<table>
  <tr>
    <th scope="col">Projects</th>
    <th scope="col">Progress</th>
  </tr>
  <tr>
    <td headers="projects100">My Project</td>
    <td>15%</td>
  </tr>
</table>
```
**IE + JAWS**: Works
**Chrome + JAWS**: Works
**Firefox + NVDA**: Works
**Safari + VO**: Works

Jym77 replied:
If this passes the SC (which I don't challenge), then the rule need to be changed to take fallback headers into account.
This certainly fails the rule right now,  so if this passes the SC, the rule is bad.

And the rule in its current version is not referencing the assignation algorithm anymore…

(technically, the rule does assume that there is no fallback headers, but if this assumption is false for the vaste majority of cases, it needs to go away)

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,295 @@
+
+## Applicability
+
+This rule applies to any `headers` attribute specified on a [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) within a `table` element, where the `table` element is [visible](#visible) and [included in the accessibility tree](#included-in-the-accessibility-tree).
```
I don't think the visibility of the table actually matters here. The headers attribute isn't visible, it's only available to assistive tech, why would a table need to be visible for this rule to apply?

EmmaJP replied:
Since 1.3.1 cares about the programmatic associations matching visually conveyed associations, I think the visibility matters.

WilcoFiers replied:
Emma's convinced me, let's leave this as is.

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
````diff
@@ -0,0 +1,295 @@
+</table>
+```
+
+### Inapplicable
````
You didn't test "not included in the accessibility tree".

EmmaJP commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,295 @@
+
+Each target attribute is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is the ID of an element, that:
+
+1. has a [semantic role](#semantic-role) of [columnheader](https://www.w3.org/TR/html-aria/#index-aria-columnheader) or [rowheader](https://www.w3.org/TR/html-aria/#index-aria-rowheader), and
```
The rule is not specific about what the element is or isn't, only that it has an expected semantic role and is part of the same table. 

As an atomic rule checking a particular thing, a fail scenario would be inaccessible, while a pass scenario would be part of a wider set of checks for a SC. So I'm not seeing how the comment is still relevant.

ShadowBB replied:
That would be a problem. Wilco states that this should not be required because it will fail situations that aren't actually a WCAG issue (VoiceOver not reading the headers correctly would not constitute a WCAG failure per se).

EmmaJP commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,295 @@
+		<th id="name" colspan="2">Name</th>
+	</tr>
+	<tr>
+		<th headers="name">Firsname</th>
```
Minor typo in the cell content

EmmaJP commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,295 @@
+			<td id="header1">Project Status</td>
+		</tr>
+		<tr>
+			<td>15%</td>
```
Should this have a headers attribute? Or should the "header1" be a `<th>` with no id?

EmmaJP commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
````diff
@@ -0,0 +1,295 @@
+		<div role="cell" headers="header1">15%</div>
+	</div>
+</div>
+```
````
Will a different rule check this kind of thing? This code could be styled to look just like a table, and has all the correct attributes.

WilcoFiers replied:
Yes, it will be tested by a different rule. The headers attribute does not work on div. It can only be on td / th. This should be in the description.

jkodu commented:
I worked with a screen reader user to test the examples on here as well as some tables with headers attribute & found that in combinations of JAWS+Chrome & NVDA+FF the table were identified and the row data were rightly associated with the column headers.

In VoiceOver there are inconsistencies.

For inapplicable examples, as expected the table were not either identified or static content was read & having gone through the results, this seems OK.

jkodu commented:
I believe this research covers what was necessary based on the comments above. I will re-enable this rule for review/ final call & await more approvals.

ShadowBB commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,310 @@
+
+#### Inapplicable Example 5
+
+The `table` is not [included in the accessibility tree][].
```
faulty brackets around link

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,351 @@
+
+Each target attribute is [a set of space separated IDs][], each of which is the ID of an element, that:
+
+1. is a [`cell`][] of the same [`table`][] as the target element.
```
Why is this a list of 1? I don't think this should be a list if it has only 1 item.

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,351 @@
+[included in the accessibility tree]: #included-in-the-accessibility-tree 'Definition of included in the accessibility tree'
+[visible]: #visible 'Definition of visible'
+[decorative]: #marked-as-decorative 'Definition of marked as decorative'
+[`cell`]: https://www.w3.org/TR/html52/tabular-data.html#cell 'Definition of cell'
```
Use whatwg here

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,351 @@
+[decorative]: #marked-as-decorative 'Definition of marked as decorative'
+[`cell`]: https://www.w3.org/TR/html52/tabular-data.html#cell 'Definition of cell'
+[semantic role]: #semantic-role 'Definition of semantic role'
+[a set of space separated ids]: https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens 'Space separated tokens'
```
why is this HTML 5.0? Use whatwg

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,351 @@
+[`cell`]: https://www.w3.org/TR/html52/tabular-data.html#cell 'Definition of cell'
+[semantic role]: #semantic-role 'Definition of semantic role'
+[a set of space separated ids]: https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens 'Space separated tokens'
+[`table`]: https://www.w3.org/TR/html50/tabular-data.html#concept-table 'Definition of table'
```
As above

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,351 @@
+
+## Accessibility Support
+
+The `headers` attribute is not supported by some popular assistive technologies as a way to assign headers in a table. Passing this rule can still cause issues for users of those assistive technologies.
```
The test data @jkodu gave me a little while ago showed that the `headers` attribute works in all major AT. Can you name me a popular browser / AT combination that doesn't work with the `headers` attribute?

jkodu replied:
Agreed. For reference here is the test data - https://github.com/act-rules/act-rules.github.io/wiki/Assistive-technology-tests-on-table-header-associations

WilcoFiers commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,351 @@
+		<th id="e2" headers="header1">2</th>
+	</tr>
+	<tr>
+		<td colspan="2" headers="header1 e1 NOT-EXIST">15%</td>
```
Just like the previous example, this table is perfectly accessible. The first assumption is true for this test case. This example needs to actually show a case where the assumption isn't true, for example, mistype the value in the `headers` attribute on the `th`.

Jym77 replied:
I don't think we should have examples where some assumptions are **false** (these would be examples where the rule cannot work, so its result is not really important).

Again, this does fail the rule in its current state. If this passes the SC (again, not challenging the fact), the rule needs to be changed.

Jym77 commented on _rules/table-headers-attribute-refer-to-data-cells-a25f45.md:
```diff
@@ -0,0 +1,351 @@
+## Assumptions
+
+- This rule assumes that the `headers` attribute is only used to identify table headers. If other information is included in the `headers` attribute, the rule may fail on issues that are not accessibility concerns. For example, if `headers` is used to include information for scripts, this rule may not be accurate.
+- This rule assumes that the `headers` attribute is required to express the relationship between data and header cells in the `table`. If the browser [computes an adequate fallback header][], this rule may produce incorrect results.
```
If this assumption is false for most cases (https://github.com/act-rules/act-rules.github.io/pull/412/files?file-filters%5B%5D=.md#r378811922) it needs to go away, and the rule updated.

jkodu commented:
Given the changes to the examples, I am adding another 1 week to this PR for final call, which will end on 24th Feb 2020
